cloudfour/cloudfour.com-patterns
# Add a Draft Email prototype

## Overview

This PR adds a "Draft Email" prototype to save the work I've been doing in WordPress land. The intent is for all of the `<textarea>` + `<button>` HTML + CSS + JavaScript logic to live inside a Gutenberg Custom HTML Block in WordPress. The CSS + JavaScript will be inlined in the Gutenberg block.

### Review Expectations

Even though this is a prototype, I would like a review of the following:
- HTML markup (you can ignore the `o-container` and footer markup, which was only copied/pasted for demo purposes)
- The `runProposedInlineJS()` function JavaScript logic (this is the JS that will be inlined in the Gutenberg Custom HTML block alongside the `<textarea>` and `<button>` elements)
- The CSS (this will also be inlined in the Gutenberg Custom HTML block)

## Screenshots

<img width="1053" alt="Screen Shot 2023-01-04 at 10 33 23 AM" src="https://user-images.githubusercontent.com/459757/210625295-69e21e6c-7d37-4fac-b418-02f7a1cbb5b3.png">

### In action

![in-action](https://user-images.githubusercontent.com/459757/211313728-d2adc7a8-3fed-441f-b7ab-8c7450ae52ba.gif)


### Clipboard API copy fail use case

![failing](https://user-images.githubusercontent.com/459757/211314735-5dafe746-b8ea-4288-97c3-5da3ea65a2c2.gif)


### Clipboard API unavailable use case

The "Copy" button is hidden when the Clipboard API is unavailable:
<img width="1084" alt="Screen Shot 2023-01-04 at 4 36 58 PM" src="https://user-images.githubusercontent.com/459757/210675907-99e7409b-c9be-4016-ae53-7d222bd2e9e4.png">

### Respecting users' motion preferences

The increased motion animation (vertical shift) only occurs when the user has no motion preference.

![reduced-motion](https://user-images.githubusercontent.com/459757/211825625-6502b101-d150-46da-82c9-3be6676cadc7.gif)


## Testing

Preview: https://deploy-preview-2109--cloudfour-patterns.netlify.app/iframe.html?args=&id=prototypes-draft-email--demo&viewMode=story

- Make sure to review the CSS animation at small and large viewports
- Make sure to also review in Safari (I had to fix a CSS layout issue with the tooltip)

### Motion preferences

- [ ] Change your motion preferences to "Reduce motion" and confirm the intro animation does not move vertically

### "Success" use case

Confirm the following:
- [ ] After activating the "Copy" button:
    - you should see a "✔️ Copied to clipboard" tooltip animate in below the "Copy" button
    - the tooltip should animate out after a few seconds
- [ ] The "Copy" button copies all of the email body text to your clipboard
- [ ] Using VoiceOver, you hear "Copied to clipboard" after activating the "Copy" button
- [ ] The "Draft email" button opens a new email with subject and body content

### "Fail" use case

Paste the following JS into your browser console:
```js
navigator.clipboard.writeText = () =>
  new Promise((resolve, reject) => {
    reject(new Error('Oopsie'));
  });
```

Confirm the following:
- [ ] After activating the "Copy" button:
    - you should see a "Whoops, failed to copy" tooltip below the "Copy" button
    - the tooltip should animate out after a few seconds
- [ ] Using VoiceOver, you hear "Whoops, failed to copy" after activating the "Copy" button
- [ ] The "Draft email" button opens a new email with subject and body content (should still work)

---

- https://github.com/cloudfour/cloudfour.com-wp/issues/945
